### PR TITLE
add MapZAP build process

### LINKED ISSUE #38: add process to assign geographies to zap records
- exploratory process has already been done in BigQuery via sql queries saved as "Project queries" and tables saved in a new `dcp_mapzap` dataset
- final table has been mapped and inspected via google's Looker Studio (free BI dashboard tool)
- would like to have a build process for this rather than rely on manually run multiple saved queries in sequence

## todo
- [x] share exploratory output with DE for review and feedback
- [x] consider the value of and good approach to moving the build process from saved BigQuery queries to this repo
- [x] do it

### PR DESCRIPTION
**Pull request by damonmcc**
resolves #38 

## changes
- add dbt project to build new MapZAP dataset
- add `.pre-commit-config.yaml` to lint dbt files
- add the docker image `google-cloud-cli` as a layer to the dev container
- update `pyproject.toml` and dev bash scripts
- update readme to describe new and pervious repo purposes

## tests
- CI tests on this PR
- locally via `pre-commit run --all-files` and `dbt test`
- inspected and shared output in BigQuery

## PR review todos
- [x] remove unnecessary uses of  `ST_ASTEXT`
- [x] remove unnecessary uses of  `COALESCE`
- [x] replace `.*` lines with explicit column names
- [x] better aliasing of source tables at the tops of models
- [x] replace `product` with `mart`

## notes
- MapZAP was initially intended to be joined to Housing DB for CEQR Type II analysis. But based on feedback from other teams, the use cases for MapZAP are more general 
- The two major issues with the current output are:
  - records are limited to the subset used for the Open Data release
  - fields are limited to the subset used for the Open Data release
- Those issues cannot be resolved until the logic in this repo which generates the Open Data release is made more modular/simplified to allow for an export of recoded data before any subsetting.
- Follow-up work will also include building out the github action `build_mapzap.yml` to allow non-local building and running a test build as part of the PR tests.
- Current lineage graph from dbt docs:
  image

### CODE REVIEW

File: bash/dev_compile_python_packages.sh [hunk 1]
```diff
@@ -5,6 +5,7 @@ set -e
 RELATIVE_SCRIPTPATH=$0
 
 # Update and install packages used to compile requirements
+apt-get -yq install python3-venv
```
Comment (fvankrieken) [inline bash/dev_compile_python_packages.sh:8]: Why is this needed?

File: dbt_project.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+clean-targets:
+  - "target"
+
+seeds:
+  +quote_columns: false
+
+models:
+  dbt_mapzap:
+    staging:
+      +materialized: view
```
Comment (fvankrieken) [inline dbt_project.yml:21]: So I was confused by their language here so I did some reading just to make sure, and want to clarify intended usage here.

I was confused because in postgres, there are materialized and non-materialized views, the former of which is stored (as a table essentially) and the latter which is just essentially a stored query. dbt uses materialized just to specify what sort of output there is, and the view here is NOT materialized. Which could be fine - less data actually being stored during running. But depending on downstream queries (and if a model needs to be queried multiple times) it could be better to store as a table

File: mapzap/models/product/product_mapzap.sql [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        project_details.*,
+        project_geometries.map_amendment_ulurp_numner,
+        project_geometries.map_amendment_project_name,
+        project_geometries.pluto_version,
+        project_geometries.bbl_areas_sum,
+        project_geometries.project_bbls,
+        COALESCE(
+            project_geometries.map_amendment_wkt is not null,
+            false
+        ) as has_map_amendment_geometry,
```
Comment (fvankrieken) [inline mapzap/models/product/product_mapzap.sql:69]: Don't think coalesce is needed here - two possibilities for `map_amendment_wki` are `null` or not `null`, and either way `is not null` will return a non-null boolean

File: mapzap/models/product/product_mapzap.sql [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+mapzap as (
+    select
+        *,
+        ST_AREA(ST_GEOGFROMTEXT(wkt)) as project_area
+    from
+        project_geometries_resolved
+)
+
+select * from mapzap
```
Comment (fvankrieken) [inline mapzap/models/product/product_mapzap.sql:100]: dbt question - is there a reason you wouldn't just end with the above select statement rather than this final `select *`?

File: mapzap/models/product/product_mapzap.sql [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        pluto_version,
+        bbl_areas_sum,
+        NULLIF(ARRAY_TO_STRING(project_bbls_array, '|'), '') as project_bbls,
+        ST_ASTEXT(project_geometry_wkt) as bbls_wkt
+    from project_bbl_geometries_aggregated
+),
+
+project_geometries as (
+    select
+        project_geometries_from_bbls.*,
```
Comment (fvankrieken) [inline mapzap/models/product/product_mapzap.sql:44]: Personal note - I really dislike `select *` statements. One in complex/long queries it can become a chore to come in and see what columns intermediate ctes actually have, and two it can make debugging hard if a source dataset has a missing column or something else unexpected (or potentially new and unwanted columns can make it through our pipeline). The first point stands regardless, but I imagine dbt helps with the second point?

File: dbt_project.yml (hunk 2 above)
Comment (damonmcc) [reply dbt_project.yml:21]: that sounds right, I definitely didn't learn enough about the pros/cons to optimize the choices between view and table here

going off of their general recommendations, the only non-view setting I made was for the `product` model, since it's the "big final table" and I imagined many queries on it to see and check the final results

File: mapzap/models/staging/base/base_dcp__pluto_with_bbls.sql [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+{% for pluto_table in pluto_versions %}
+
+    SELECT
+        '{{ pluto_table }}' AS pluto_version,
+        CAST(bbl AS STRING) AS bbl,
+        wkt
+    FROM {{ source('map_pluto', pluto_table) }}
+
+    {% if not loop.last %}
+        UNION ALL
```
Comment (fvankrieken) [inline mapzap/models/staging/base/base_dcp__pluto_with_bbls.sql:30]: So this is unioning all versions of pluto? And what's the rationale for the two pluto source tables?

File: mapzap/models/product/product_mapzap.sql (hunk 3 above)
Comment (damonmcc) [reply mapzap/models/product/product_mapzap.sql:69]: ahhh I see

so just `project_geometries.map_amendment_wkt is not null` should do it?

Comment (fvankrieken) [reply mapzap/models/product/product_mapzap.sql:69]: Yup

File: dbt_project.yml (hunk 2 above)
Comment (fvankrieken) [reply dbt_project.yml:21]: Sounds good. I think as a rule defaulting to views until you run into performance issues makes sense (or if you know you have one model that's going to be referenced a lot within the build for some reason)

File: mapzap/models/product/product_mapzap.sql (hunk 4 above)
Comment (damonmcc) [reply mapzap/models/product/product_mapzap.sql:100]: no functional reason, I think it's just a convention for readability/simplifying of CTEs. I sometimes do `order by` in the final `select *` and it may be a nice place to select only certain columns instead of `*`

[dbt CTE docs](https://docs.getdbt.com/terms/cte)

File: mapzap/models/product/product_mapzap.sql [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        pluto_version
+),
+
+project_geometries_from_bbls as (
+    select
+        project_id,
+        pluto_version,
+        bbl_areas_sum,
+        NULLIF(ARRAY_TO_STRING(project_bbls_array, '|'), '') as project_bbls,
+        ST_ASTEXT(project_geometry_wkt) as bbls_wkt
```
[the remark this answers is not in the document]
Comment (damonmcc) [reply mapzap/models/product/product_mapzap.sql:38]: I think this was overkill to make sure this exported to csv/shapefiles well and I probably blamed an "invalid geometry" issue on this before I fixed it elsewhere.

I'll drop these and see if it all works without em

File: mapzap/models/staging/base/base_dcp__pluto_with_bbls.sql (hunk 6 above)
Comment (damonmcc) [reply mapzap/models/staging/base/base_dcp__pluto_with_bbls.sql:30]: this is unioning all versions that already have BBLs

the two pluto base tables are for different types of pluto versions: those that already have BBLs and those that need BBLs constructed. and then the staging table `stg_dcp__pluto_bbls` combines them for downstream use

the seed file `seed_pluto_versions.csv` (which becomes a table via `dbt seed`) stores the details for each pluto version

File: mapzap/models/staging/base/base_dcp__pluto_with_bbls.sql [hunk 8]
```diff
@@ -0,0 +1,32 @@
+{% set pluto_versions_query %}
```
Comment (fvankrieken) [inline mapzap/models/staging/base/base_dcp__pluto_with_bbls.sql:1]: What's this line do? Where is it actually getting this query from?

Interesting file btw, cool to see the jinja iteration in action

**Comment by damonmcc** [comment]
> What is needed in `.env` to run this locally?
> 
> Edit - saw example.env, sorry

nice. also I made a change that might be more of a pain than it's worth.

I started putting `.env` in the dev container folder to make it more clear that it's only used locally via dev container. I also added `runArgs` to `devcontainer.json` to automatically load the env file.

the double-edge takeaway is that we don't need to keep reloading environment variables at the top of every bash script (this build doesn't use one). the con being that a change to `.env` doesn't take effect unless the dev container is rebuild or it's manually done in a terminal/script

File: mapzap/models/staging/base/base_dcp__pluto_with_bbls.sql (hunk 6 above)
Comment (fvankrieken) [reply mapzap/models/staging/base/base_dcp__pluto_with_bbls.sql:30]: Ohh so the has bbl vs needs bbl constructed is not a record-level issue but a whole table (version of pluto) level issue.

Comment (fvankrieken) [reply mapzap/models/staging/base/base_dcp__pluto_with_bbls.sql:30]: As for unioning all versions (in the "final" staging pluto source table), how are duplicate bbls (with potentially differing geometries) dealt with? Just unioning all geometries in the aggregate step, right? Do we have any concern with using potentially outdated info? It seems like it should be fine, just sort of thinking out loud here

**Comment by fvankrieken** [comment]
> > What is needed in `.env` to run this locally?
> > Edit - saw example.env, sorry
> 
> nice. also I made a change that might be more of a pain than it's worth.
> 
> I started putting `.env` in the dev container folder to make it more clear that it's only used locally via dev container. I also added `runArgs` to `devcontainer.json` to automatically load the env file.
> 
> the double-edge takeaway is that we don't need to keep reloading environment variables at the top of every bash script (this build doesn't use one). the con being that a change to `.env` doesn't take effect unless the dev container is rebuild or it's manually done in a terminal/script

I like that, I think that's cleaner than loading it in bash/python

File: bash/dev_compile_python_packages.sh (hunk 1 above)
Comment (damonmcc) [reply bash/dev_compile_python_packages.sh:8]: this became necessary after I changed the dev container Dockerfile to use the `gcr.io/google.com/cloudsdktool/google-cloud-cli` image. after some messing around, this seemed the easiest/most reliable way to get the google cloud SDK in the dev container (it's [used in GHAs](https://github.com/NYCPlanning/db-zap-opendata/blob/build-map-zap/.github/workflows/export.yml#L57-L62))

I didn't mess around too much with optimizing the use of that new image layer and our usual `mcr.microsoft.com/devcontainers/python` image layer, but it only worked when the google one was last, and we _may_ not need the old one at all

note: I ended up not _needing_ the google cloud SDK, but since it's used in other parts of the repo in CI and would be needed for any auto exports/imports from bigquery to file storage, it seems nice to have ready to use locally

Comment (fvankrieken) [reply bash/dev_compile_python_packages.sh:8]: Gotcha, that makes sense

File: mapzap/models/product/product_mapzap.sql (hunk 5 above)
Comment (fvankrieken) [reply mapzap/models/product/product_mapzap.sql:44]: Edited to tone down my strong language!

File: mapzap/models/summary/summary_mapzap_bbls.sql [hunk 9]
```diff
@@ -0,0 +1,57 @@
+with project_geometries as (
+    select * from {{ ref('product_mapzap') }}
```
Comment (fvankrieken) [inline mapzap/models/summary/summary_mapzap_bbls.sql:2]: I feel like less extra aliases (just referencing source tables as `{{ ref(...) }}` in the queries below) could keep it cleaner as to where data is coming from and how it's being used in the actual meat of the sql file. I take it you prefer this aliasing?

File: mapzap/models/staging/base/base_dcp__pluto_with_bbls.sql (hunk 8 above)
Comment (fvankrieken) [reply mapzap/models/staging/base/base_dcp__pluto_with_bbls.sql:1]: Gotcha. Didn't realize that was valid syntax, but also should have just read a little more closely

File: mapzap/models/staging/base/base_dcp__pluto_with_bbls.sql (hunk 6 above)
Comment (damonmcc) [reply mapzap/models/staging/base/base_dcp__pluto_with_bbls.sql:30]: ooo there's [a test for that](https://github.com/NYCPlanning/db-zap-opendata/blob/build-map-zap/mapzap/models/staging/_dcp_models.yml#L124-L127)!

duplicate BBLs in the same pluto version would be bad, so that test checks for that. but a BBL existing and/or having different geometry in a different pluto version is ok since a ZAP record only looks at a single year (aka pluto version) to get BBL geometries

File: mapzap/models/summary/summary_mapzap_bbls.sql (hunk 9 above)
Comment (fvankrieken) [reply mapzap/models/summary/summary_mapzap_bbls.sql:2]: That's a good point that you get a clean summary of the sources, I don't know if it quite wins me over though lol. I'm in the middle. Wouldn't a couple of comments listing sources get the job done in fewer lines for that purpose?

I think your last point is good too - if we go this route, just keep the exact names to avoid confusion.

File: mapzap/models/staging/stg_dcp__zap_bbls.sql [hunk 10]
```diff
@@ -0,0 +1,21 @@
+with source_zap_project_bbls as (
+
+    select * from {{ source('zap_bbls', '20230526') }}
```
Comment (fvankrieken) [inline mapzap/models/staging/stg_dcp__zap_bbls.sql:3]: What's producing this table? Other code in this repo?

Comment (fvankrieken) [reply mapzap/models/staging/stg_dcp__zap_bbls.sql:3]: Sorry you're getting a lot of just general knowledge questions and not just code review! But just trying to understand what specifically is happening in all of this

**Comment by fvankrieken** [comment]
I do wonder if maybe we should just stick with dbt's suggested names. I don't love staging/intermediate/mart but at least someone can google and find documentation. Or even I feel like product is clear enough that this current set of labelling works

**Comment by fvankrieken** [comment]
Besides outstanding comments, lgtm. And even just giving things another lookover now, beginning to warm up to dbt a tad more :)

File: mapzap/models/staging/stg_dcp__zap_bbls.sql (hunk 10 above)
Comment (damonmcc) [reply mapzap/models/staging/stg_dcp__zap_bbls.sql:3]: haha worries! this table is a source that's expected to exist already for the build to use (the result of EL before the T)

all source tables are declared in [`_dcp_source.yml`](https://github.com/NYCPlanning/db-zap-opendata/blob/build-map-zap/mapzap/models/staging/_dcp_source.yml) and (by convention) are references using `{{ source(name, table) }}` rather than `{{ ref ... }}`

as I'm typing this, I think the advantages to using `{{ source ... }}` are that the metadata then knows it's a source and not a model and if it isn't configured in `_dcp_source.yml`, it can't be found and has an informative error

Comment (fvankrieken) [reply mapzap/models/staging/stg_dcp__zap_bbls.sql:3]: I actually mean specifically how is this source generated? Are we producing it? The bigquery table

Comment (damonmcc) [reply mapzap/models/staging/stg_dcp__zap_bbls.sql:3]: ah, yea we produce it via this repo's weekly export of ZAP CRM data that I improved in https://github.com/NYCPlanning/db-zap-opendata/pull/42 last week

File: mapzap/models/product/product_mapzap.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (damonmcc) [reply mapzap/models/product/product_mapzap.sql]: fixed in https://github.com/NYCPlanning/db-zap-opendata/pull/40/commits/1a9509505960d29fe157118b86d6fcd852438d62

File: mapzap/models/summary/summary_mapzap_bbls.sql (hunk 9 above)
Comment (damonmcc) [reply mapzap/models/summary/summary_mapzap_bbls.sql:2]: I made the "exact name" change in the summary models with https://github.com/NYCPlanning/db-zap-opendata/pull/40/commits/0c4867ca709223baf909268bf0b81e1165b74f71

as I started to do it to the product and intermediate models though, I hit what feels like the main value of an alias: to address tables with a shorter or more meaningful name AND decouple the code from the table name

if a source table name changes, we can either:
- use an alias and change one line at the top of the file
- not use an alias and have to change multiple lines of code and comments

**Comment by damonmcc** [comment]
@fvankrieken

agreed on the value of using staging/intermediate/marts. I changed product to marts

File: mapzap/models/summary/summary_mapzap_bbls.sql (hunk 9 above)
Comment (fvankrieken) [reply mapzap/models/summary/summary_mapzap_bbls.sql]: Hmm. Alright, that seems fine for now, but definitely this a norm to keep thinking about. We might have made a compromise that leaves neither of us happy lol

**Review by fvankrieken — APPROVED** [review]
